design/datamover.md

```diff
@@ -0,0 +1,208 @@
+It is desired that Velero support a concept of Datamover. We would like to propose an extensible design to support Datamovers, where-in the users
+can bring their own datamover implementation and use it to replicate CSI snapshots to object storage during backup and also recreate PVCs from the snapshots
+in object storage during restore. Thus making the CSI snapshots durable as well as portable.
+
+## Background
+Currently, Velero has two option to support backing up of persistent volumes.
+
+1. Restic: This solution performs filesystem level backups and mostly used when a volume snapshot plugin is not available. However, the restic integration is
+   tightly coupled with Velero, makes it difficult for someone to bring their own implementation of data movement. Although restic is portable but it is not
+   crash consistent as it requires quiescing of data before the backup is taken thus affecting the availability of application.
+
+2. Volume Snapshots: Storage providers provide with volume snapshot capabilities via plugins. These snapshots are crash consistent but there is no guarantee that
+   these snapshots are durable. These snapshots are generally stored locally in the cluster storage, so if the cluster breaks then the snapshots are of no use. Velero also supports
+   CSI volume snapshots, but as there is no data movement provided by Velero for these CSI snapshots they are not durable and dependent on the storage provider.
+
+
+Related Issue: https://github.com/vmware-tanzu/velero/issues/3229
+
+## Goals
+- To be able to backup CSI snapshots to object storage
+- To be able to recreate CSI snapshots under a different storage provider
+- To provide an extensible data mover solution which should allow third parties to plugin their own datamover implementation
+- To provide a Datamover example PoC
+
+## Non Goals
+- Maintain 3rd party data mover implementations
+- Adding a status watch controller to Velero
+- To support incremental backups (future scope, dependent on standard CSI support for snapshot differentials)
+- To allow plug-ins for services like encryption, compression etc. (future scope)
+
+## High-Level Design ([VolSync](https://github.com/backube/volsync) as a working Proof of Concept)
+The goal of this high level design is to explain a working PoC of moving CSI snapshots where-in a datamover (VolSync) works in conjunction with Velero via a DataMover plugin (will be installed when the DataMover flag is set to true).
+
+The following design proposal is for adding a datamover feature to Velero which can be integrated with various user implemented datamovers.
+
+![](datamover.png)
+
+The design components and CRs that we are introducing are as follows:
+
+1. DataMoverType(DMT): This is a cluster scoped Custom Resource that will have details about the data mover provisioner.The provisioner will be responsible
```

Comment:
I understand the idea that a DMT maps to an actual mover that moves the data.
But the workflow is a little confusing to me, could you clarify:
Is this `DataMover plugin` a velero plugin ?  
How does it also responsible for installing the controllers?
How the plugin creates the CR.

Reply:
@reasonerjt So there will be a BackupItemAction plugin which creates DataMoverBackup resources in the cluster upon backing up a PVC or CSI snapshot, and a RestoreItemACtion plugin which creates a DataMoverRestore upon restoring a DataMoverBackup. These would just be normal velero plugins and would not be responsible for installing controllers.

Reply:
@sseago 
A little curious about whether the plugin will do the data movement itself, or the plugin will also call some interface implemented by other components?